### ISSUE TITLE
Trait resolution fails with incorrect error when using functions as input parameters

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust#62385]
This works ([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e0fd0844e20c87b6e28c8e246aad154c))
:

```rust
trait Call<F> {
    type Ret;
    fn call(self, f: F) -> Self::Ret;
}

impl Call<f32> for f32 {
    type Ret = f32;
    fn call(self, f: f32) -> Self::Ret {
        f
    }
}
impl Call<i32> for f32 {
    type Ret = i32;
    fn call(self, f: i32) -> Self::Ret {
        f
    }
}

fn main() {
    let _ = (0_f32).call(0_i32);
}

```

but this fails ([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=eb1c03e73ae9b722fb8af8b8f60482cb)):

```rust
trait Call<F> {
    type Ret;
    fn call(self, f: F) -> Self::Ret;
}

impl Call<fn(f32) -> f32> for f32 {
    type Ret = f32;
    fn call(self, f: fn(f32) -> f32) -> Self::Ret {
        f(self)
    }
}
impl Call<fn(f32) -> i32> for f32 {
    type Ret = i32;
    fn call(self, f: fn(f32) -> i32) -> Self::Ret {
        f(self)
    }
}

fn main() {
    fn u0(x: f32) -> f32 {
        x
    }
    let _ = (0_f32).call(u0);
}

```

with

```
error[E0277]: the trait bound `f32: Call<fn(f32) -> f32 {main::u0}>` is not satisfied
  --> src/main.rs:23:21
   |
23 |     let _ = (0_f32).call(u0);
   |                     ^^^^ the trait `Call<fn(f32) -> f32 {main::u0}>` is not implemented for `f32`
   |
   = help: the following implementations were found:
             <f32 as Call<fn(f32) -> f32>>
             <f32 as Call<fn(f32) -> i32>>

error: aborting due to previous error
```

I'm not sure if this is by design or not (AFAICT the trait bound is satisfied), but the error message could be at least much clearer.

**Comment by jonas-schievink** [comment]
This is expected, since `u0` doesn't have type `fn(f32) -> f32`, since every function item gets its own anonymous type (like closures). These types will coerce to the corresponding function pointer type (`fn(f32) -> f32` here), but that coercion only kicks in when there's a known expected type to coerce to.

**Comment by gnzlbg** [comment]
Ah! So that's what's the ` {main::u0}` in the `fn(f32) -> f32 {main::u0}` is! 

I think this error could use a note, explaining that the `fn(f32) -> f32 {main::u0}` is an anonymous type, different from `fn(f32) -> f32`, but that it can be coerced into `fn(f32) -> f32`, although this coercions are not triggering here. 

Maybe suggest using `.. as fn(f32) -> f32` here to trigger the coercion. 

cc @estebank

**Comment by estebank** [comment]
Another [example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=e2fa8c3b5288b86343c7d2b6633ec21e) from [reddit](https://www.reddit.com/r/rust/comments/deeoph/hey_rustaceans_got_an_easy_question_ask_here/f2y00le/?context=8&depth=9):

```rust
struct Dummy;

impl From<fn(u32) -> String> for Dummy {
    fn from(_: fn(u32) -> String) -> Self { unimplemented!() }
}

fn foo(_: u32) -> String { unimplemented!() }

fn main() {
    Dummy::from(foo);
}
```

we currently emit

```
error[E0277]: the trait bound `Dummy: std::convert::From<fn(u32) -> std::string::String {foo}>` is not satisfied
  --> src/main.rs:10:5
   |
10 |     Dummy::from(foo);
   |     ^^^^^^^^^^^ the trait `std::convert::From<fn(u32) -> std::string::String {foo}>` is not implemented for `Dummy`
   |
   = help: the following implementations were found:
             <Dummy as std::convert::From<fn(u32) -> std::string::String>>
   = note: required by `std::convert::From::from`
```

The solution is changing the call to `Dummy::from(foo as fn(u32) -> String);`.

**Comment by benesch** [comment]
I know this issue is about improving the diagnostic, but it seemed as good a place as any to note that fixing this _is_ desirable, but apparently there are a number of implementation challenges, according to @nikomatsakis ([Zulip source](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/More.20coercions.20from.20fn.20item.20-.3E.20fn.20pointer/near/199921865)).

https://github.com/rust-lang/rust/issues/62621#issuecomment-510867961 suggested that the next step might be an RFC, but it sounds like the more pressing issue is sorting through some of the implementation challenges.

**Comment by estebank** [comment]
Current output:

Only change to the path of the `fn` item

```
error[E0277]: the trait bound `f32: Call<fn(f32) -> f32 {u0}>` is not satisfied
  --> src/main.rs:23:26
   |
23 |     let _ = (0_f32).call(u0);
   |                     ---- ^^ the trait `Call<fn(f32) -> f32 {u0}>` is not implemented for `f32`
   |                     |
   |                     required by a bound introduced by this call
   |
   = help: the following other types implement trait `Call<F>`:
             <f32 as Call<fn(f32) -> i32>>
             <f32 as Call<fn(f32) -> f32>>
```

Much better output for this case

```
error[E0277]: the trait bound `Dummy: From<fn(u32) -> String {foo}>` is not satisfied
  --> src/main.rs:10:5
   |
10 |     Dummy::from(foo);
   |     ^^^^^ the trait `From<fn(u32) -> String {foo}>` is not implemented for `Dummy`
   |
   = help: the trait `From<fn(u32) -> String>` is implemented for `Dummy`
   = help: for that trait implementation, expected `fn(u32) -> String`, found `fn(u32) -> String {foo}`
```

We need to explicitly check for this case and detect `fn(u32) -> String { foo }` can be cast to `fn(u32) -> String` and provide a structured suggestion to do so for both cases presented.
